### ISSUE TITLE
Rocket Chat is not fully accessible for Screen Readers

### ISSUE THREAD

**Issue by carlosandrews** [RocketChat/Rocket.Chat#14814]
### Description:
Users who use Job Access With Speech (JAWS) are not fully able to use Rocket Chat when accessed from a web browser only.

Recommend testing with JAWS/NVDA or use tools like Accessible Name & Description Inspector ([ANDI](https://www.ssa.gov/accessibility/andi/help/install.html)).

Testing with JAWS:
https://webaim.org/articles/jaws/

Testing with NVDA:
https://webaim.org/articles/nvda/

DHS Harmonized process for Section 508 testing of web applications:
https://www.dhs.gov/sites/default/files/publications/Baseline_Tests_for_Software_and_Web_Accessibility_v2_0_2.pdf

DHS Section 508 Compliance Testing Tools:
https://www.dhs.gov/508-tools

Accessibility testing tools from The Paciello Group:
https://developer.paciellogroup.com/blog/2019/02/accessibility-testing-tools-we-use/

### Steps to reproduce:

1. Open a Rocket Chat instance from a web browser (e.g. Chrome)
2. Utilize the [ANDI](https://www.ssa.gov/accessibility/andi/help/install.html) bookmarklet
3. Download [NVDA](https://www.nvaccess.org/download/) (open source screen reader)
4. Test Rocket Chat as per above reference with NVDA
5. If needed, purchase [JAWS](https://www.freedomscientific.com/products/software/jaws/)
6. Test with JAWS as per above reference
7. Use the DHS Harmonized process for Section 508 testing of web applications as a reference.

### Server Setup Information:

- Version of Rocket.Chat Server: 
- Operating System: Windows 10 64-bit
- Deployment Method: 
- Number of Running Instances: 
- DB Replicaset Oplog: 
- NodeJS Version: 
- MongoDB Version: 
- Browser version: Google Chrome Version 74.0.3729.131 (Official Build) (64-bit)

**Comment by Sing-Li** [comment]
@carlosandrews  please DM me on open if you are available.

**Comment by carlosandrews** [comment]
> @carlosandrews please DM me on open if you are available.

How do I do that?

**Comment by carlosandrews** [comment]
> @carlosandrews please DM me on open if you are available.

@Sing-Li How do I do that?

**Comment by ethindp** [comment]
I am also having accessibility problems of my own on my instance of rocket.chat. For example, if I press the "more" button in chat rooms, the menu doesn't display a normal dropdown like most menus that I've used on the web do. When I navigate down to the menu, it lets me navigate among a few options and then vanishes. In order for me to make it remain open, I have to explicitly navigate by list (which could be challenging if I'm in a very large discussion/room with lots of messages). Also, I can't create a public channel either. I click on 'create new', select 'channel' (with the same menu behavior as explained above) and it doesn't show public anywhere (and I have to guess as to which option I'm selecting when I just press enter on the channel type I want). Public, in fact, is shown nowhere. If I create the channel and then try setting its type to public, it doesn't let me save it and makes the save button unavailable, and if I do set it to a non-privat type it still lists the channel under 'Private Groups'.

**Comment by mgifford** [comment]
I suspect that this is too broad a way to approach this. 

I'd love to know if Rocket.Chat works well for keyboard only users.  I haven't done any testing on this, but so many web interfaces depend on mice. 

It would also be useful to know if automated tools like:
  https://accessibilityinsights.io/
  http://wave.webaim.org/

identify problems.  These tools are easy for developers to use and evaluate. 

Next thing is to see if you can get it working in either NVDA or VoiceOver as a sighted person.  If your a developer, it should be pretty easy to install NVDA or windows or just learn how to use VoiceOver.  

Then you can start engaging with blind users. There are a bunch that are already using Rocket.Chat! You might already know who they are.  If there is any $$ you might engage with groups like https://www.makeitfable.com/

I'm also a big fan of accessibility statements:
https://github.com/accessibility/Accessibility-Statement

But to address accessibility issues in software it is important to break it down into little parts that can be tested and approved by developers.  Problems have to be specific and you don't just start with testing on JAWS. I'm an accessibility professional and I know that I do not know enough to evaluate a site as a blind user does.

**Comment by valeriemichaud-k1** [comment]
Just checking if there is any accessibility roadmap?

**Comment by echuber2** [comment]
Could this issue be reopened? Does anyone have recent experience with improvements or changes this year on the accessibility front?

**Comment by ZathrasXI** [comment]
Have just tried using Rocket Chat in Safari with VoiceOver and found it very confusing at times. For example, to access the sub-menus for: the avatar button, the filters button and the create new channel button, you need to activate the button and then navigate through all of the other elements on the page before arriving on the menu, but VoiceOver didn't even tell me that I was on the menu when I eventually arrived on it :(

**Comment by ZathrasXI** [comment]
Also, it seems like screen reader (VoiceOver) is not notified of new messages. For example, when I am in a chat and receive a new message then the screen reader does not announce that I have received a new message.

**Comment by ekavun** [comment]
I sort of dropped the ball on this. I just updated to 3.18.0 and some accessibility issues persist. Running Mac and tested with Voiceover on Safari and Chrome.

* Buttons such as Directory, Display, etc. will read once, then fail to be recognized with the text label. 
* It is still impossible to attach files. Copy paste does not work, and there is no 'slash' command to trigger the file-upload popup. This is the most blocking issue for me.
* As an improvement, Live Regions can be used to automatically speak incoming messages. 
Shall we open or re-open a #Accessibility channel and get some commitment from impacted users to test and provide improvements? Happy to contribute wherever I can.

**Comment by ekavun** [comment]
To be clear, the icon to trigger the file-upload button is not visible, nor clickable with VoiceOver or Jaws.

**Comment by Malvoz** [comment]
> Also, it seems like screen reader (VoiceOver) is not notified of new messages.

> As an improvement, Live Regions can be used to automatically speak incoming messages.

While [Live Regions](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions) can be achieved using relevant ARIA attributes such as [`aria-live`](https://www.w3.org/TR/wai-aria/#aria-live), [`aria-atomic`](https://www.w3.org/TR/wai-aria/#aria-atomic) and [`aria-relevant`](https://www.w3.org/TR/wai-aria/#aria-relevant), it should be noted that [`role="log"`](https://www.w3.org/TR/wai-aria/#log) is a specialized ARIA role for messages, and is implicitly a Live Region. [Twitch](https://twitch.tv) uses `role="log"` on the messages' container element, and as such new messages are announced to screen reader users.

**Comment by ivans-netto** [comment]
Hi all 👋 Ivan here, Head of Design at Rocket.Chat.

Screen reader users frequently encounter challenges at some points within Rocket.Chat because the platform is continually adjusting its content to enhance accessibility for JAWS/NVDA and VoiceOver users.

We've made various improvements on accessibility and specially while navigating Rocket.Chat platform using assistive technologies. You can check the new accessibility features opening the user menu in your sidebar > accessibility and appearance; there you're gonna find themes and other accessibility features.

**Our accessibility program is currently in the process of implementing a more advanced user experience with JAWS, NVDA, and VoiceOver. The expected release date for these enhancements is March 31, 2024.**

If you'd like to collaborate with us bringing up issues with screen readers, don't hesitate on sending me a DM to me (@ivan.netto) on open.rocket.chat.

**Comment by dougfabris** [comment]
Hello everyone, as @ivans-netto shared we've been working on a bunch of improvements this year and we'll continue improving in the next years. Many of the issues you shared we already covered and I would like to share some news with you.

@ethindp Thanks for your consideration. About the more button, we replaced these menus with brand new ones:

![Kapture 2023-12-21 at 14 40 19](https://github.com/RocketChat/Rocket.Chat/assets/27704687/2281c411-939c-4312-b059-11b092efc5a4)

We improved the fields on the Create new channel modal, so now the screen reader reads the private switch option.
Also, we improved the channel settings as well but we still have an issue with focus (we're working on that).

![Kapture 2023-12-21 at 14 37 53](https://github.com/RocketChat/Rocket.Chat/assets/27704687/db1c93cc-b89a-4803-9105-392a306f0c6b)
 
@mgifford Thanks for sharing the tools and the groups, Fable seems like a great opportunity to help us improve our product.  We’re working hard to reduce the amount of a11y issues, it's working better but we do have a bunch of changes to do yet. I would like to invite you if you have the opportunity to test again and send some feedback are more than welcome

@valeriemichaud-k1 We worked on a bunch of issues this year and we achieved a [BITV certification](https://www.rocket.chat/blog/pioneering-digital-accessibility-with-bitv-2-0-certification), for the next year, we'll be focused on improving even more our accessibility. Unfortunately, the roadmap is not public.

@zathrasXI - We worked on the avatar button and those menus as you can see on the gifs I attached above. We're working on the message room to improve the experience when receiving new messages.

@ekavun - We worked on those buttons, and we rewrote the entire composer, and now is possible to attach files using the screen reader

![Kapture 2023-12-21 at 15 01 13](https://github.com/RocketChat/Rocket.Chat/assets/27704687/8017ecf6-e40c-49ad-9530-ad2080b96418)

@malvoz Thanks for the suggestion, we’re planning to work on improving the message room very soon.

@ekavun and everyone involved here, I wanted to walk through those points because it's very important to us to have our product accessible, and as @mgifford said: 

> “But to address accessibility issues in software it is important to break it down into little parts that can be tested and approved by developers. Problems have to be specific and you don't just start with testing on JAWS. I'm an accessibility professional and I know that I do not know enough to evaluate a site as a blind user does. But to address accessibility issues in software it is important to break it down into little parts that can be tested and approved by developers. Problems have to be specific and you don't just start with testing on JAWS. I'm an accessibility professional and I know that I do not know enough to evaluate a site as a blind user does.” 

Because of that I will close this issue and will ask you to revisit our latest version, share feedback, and open issues, everything you could do to help us is more than welcome. 

Also as @ekavun suggested we opened a channel dedicated to accessibility contributions and I would like to invite everyone who would like to help us to join the channel on this link: [accessibility-program-feedback](https://go.rocket.chat/invite?host=open.rocket.chat&path=invite%2FPKoxzk)

Thank you so much for helping us be loved community ❤️
